### ISSUE TITLE
Scheduler creating an unbounded number of threads at startup

### ISSUE THREAD

**Issue by ghostdogpr** [zio/zio#8371]
_Even though based on Discord discussions there is no intention of fixing this, I'm opening an issue for people who may run into it._

We've been observing that when starting an application that immediately receives important load, the zio 2 scheduler is creating a unbounded number of threads, leading to contention at the classloader level as well as loading lazy vals.

A couple pictures first (note the number of `ZScheduler` threads: 378 😱):

<img width="793" alt="Screen Shot 2023-09-08 at 9 53 31 AM" src="https://github.com/zio/zio/assets/7413894/6b6c9d98-65d9-4dcf-ab68-4734e9fb6f71">

They're all waiting on the same locks (here, a simple lazy val):

<img width="703" alt="Screen Shot 2023-09-08 at 9 54 15 AM" src="https://github.com/zio/zio/assets/7413894/e99456f6-266b-473c-8d98-579fc644caf3">

What seems to happen is that the first access to a given class or lazy val is slow (as the JVM just started and classes are not loaded yet), so we expect a little latency. But the `ZScheduler` detects that the thread is "blocking" and decides to create another one. Since we have multiple fibers doing that at the same time, this process repeats multiple times, leading to a thread explosion. What makes it worse is that all those threads are competing for the same locks (see the second picture), so it completely freezes the JVM until those resources are finally loaded. For us it resulted in 10~20 seconds freeze in production when starting to receive load.

As a workaround, we switched to a custom executor based on zio 1 truly bounded executor (using `Runtime.setExecutor(Executor.fromThreadPoolExecutor(...))`. Because it is bounded, it does not create new threads and with very little contention all the classloading happens rather quickly (<1s).

It might be a rare use case: large application with lots of classes (which makes it difficult to "pre-load"), instant intense usage. But this behavior seems to me like it could cascade easily even in other use cases.

I think it could be nice to have the option to keep a bounded number of threads (basically disabling auto-blocking behavior). Auto-blocking might be nice for users who don't know or don't want to care about what's blocking but I would love to have the possibility to decide where my code runs and keep a bounded threadpool.

**Comment by adamgfraser** [comment]
I think this is best addressed by using  a `Semaphore` that limits the maximum number of concurrent requests served at the same time (I believe libraries like `rezilience` provide policies that do this out of the box or ZIO HTTP could provide it) or specifying a custom executor as you have done, but definitely a good watch out!

**Comment by ghostdogpr** [comment]
I don't want to limit concurrent requests because apart from that startup behavior, my application is able to handle the load just fine and there is no reason to make it slower.

By the way, apart from the auto-blocking logic, is there anything `ZScheduler` does that we're missing out by not using it?

**Comment by adamgfraser** [comment]
But you're very explicitly using a fixed thread pool as a way of limiting concurrency. In essence if you have `n` cores you are using your thread pool to make sure that you only have `n` requests being served at the same time, and are relying on this behavior at startup which is causing your problems. So I think it would be more straightforward to just do that explicitly.

**Comment by ghostdogpr** [comment]
Having n requests served concurrently is not the same as having n threads since requests handling involve other things such as waiting for I/O. A request might be still pending without using a core.

**Comment by adamgfraser** [comment]
Sure they are not exactly the same but both of them are ways to limit the concurrency of your application which is your problem here. Anyway we could look at providing a way to make this more configurable
